LeanOA/ForMathlib/Analysis/SpecialFunctions/ContinuousFunctionalCalculus/Abs.lean

```diff
@@ -0,0 +1,236 @@
+    ← cfcₙ_mul ..]
+  conv_rhs => rw [← cfcₙ_id (R := ℝ) a]
+  refine cfcₙ_congr fun x hx ↦ ?_
+  refine Real.mul_self_sqrt ?_
+  exact quasispectrum_nonneg_of_nonneg a ha x hx
+
+lemma abs_of_nonneg (a : A) (ha : 0 ≤ a := by cfc_tac) : abs a = a := by
+  rw [abs, ha.star_eq, sqrt_mul_self a ha]
+
+lemma abs_of_nonpos (a : A) (ha : a ≤ 0 := by cfc_tac) : abs a = -a := by
+  simp [← abs_neg a, abs_of_nonneg _ <| neg_nonneg.mpr ha]
+
+lemma abs_eq_norm (a : A) (ha : IsSelfAdjoint a := by cfc_tac) :
+    abs a = cfcₙ (‖·‖) a := by
+   simp only [abs, Real.norm_eq_abs, ← Real.sqrt_sq_eq_abs, sq]
+   have H : cfcₙ Real.sqrt (a * a) = cfcₙ (fun x ↦ √(x * x)) a := by
+       rw [← Function.comp_def, cfcₙ_comp a (f := fun x ↦ x * x) (g := fun x ↦ √x), cfcₙ_mul .., cfcₙ_id' ..]
+   rw [sqrt_eq_real_sqrt (ha := star_mul_self_nonneg a), ha.star_eq, H]
+
+protected lemma posPart_add_negPart (a : A) (ha : IsSelfAdjoint a := by cfc_tac) : abs a = a⁺ + a⁻ := by
+  rw [CFC.posPart_def, CFC.negPart_def, ← cfcₙ_add .., abs_eq_norm a ha]
+  exact cfcₙ_congr fun x hx ↦ (posPart_add_negPart x).symm
+
+lemma abs_sub_self (a : A) (ha : IsSelfAdjoint a := by cfc_tac) : abs a - a = 2 • a⁻ := by
+  nth_rw 2 [← CFC.posPart_sub_negPart a]
+  rw [CFC.posPart_add_negPart a]
+  abel
+
+lemma abs_add_self (a : A) (ha : IsSelfAdjoint a := by cfc_tac) : abs a + a = 2 • a⁺ := by
+  nth_rw 2 [← CFC.posPart_sub_negPart a]
+  rw [CFC.posPart_add_negPart a]
+  abel
+
+@[simp]
+lemma abs_abs (a : A) : abs (abs a) = abs a :=
+  abs_of_nonneg ..
+
+variable [StarModule ℝ A]
+
+lemma abs_smul_real (r : ℝ) (a : A) : abs (r • a) = |r| • abs a := by
+  have : 0 ≤ |r| • abs a := smul_nonneg (by positivity) abs_nonneg
+  rw [abs, CFC.sqrt_eq_iff _ _ (star_mul_self_nonneg _) this]
+  simp only [mul_smul_comm, smul_mul_assoc, abs_mul_self, star_smul]
+  match_scalars
+  simp
+
+lemma abs_smul_nnreal (r : ℝ≥0) (a : A) : abs (r • a) = r • abs a := by
+  simpa [NNReal.abs_eq] using abs_smul_real r a
+
+end Real
+
+section Complex
+
+variable [NonUnitalRing A] [TopologicalSpace A] [Module ℂ A]
+
+variable [StarRing A] [PartialOrder A] [StarOrderedRing A]
+variable [IsScalarTower ℂ A A] [SMulCommClass ℂ A A]
+variable [NonUnitalContinuousFunctionalCalculus ℂ (IsStarNormal : A → Prop)]
+
+open ComplexOrder in
+lemma cfcₙ_norm_sq_nonneg {f : ℂ → ℂ} {a : A} : 0 ≤ cfcₙ (fun z ↦ star (f z) * (f z)) a :=
+  cfcₙ_nonneg fun _ _ ↦ star_mul_self_nonneg _
+
+open ComplexOrder in
+lemma cfcₙ_norm_nonneg {a : A} : 0 ≤ cfcₙ (fun z : ℂ ↦ (‖z‖ : ℂ)) a :=
+  cfcₙ_nonneg fun _ _ ↦ by simp only [Complex.zero_le_real, norm_nonneg]
+
+variable [NonnegSpectrumClass ℝ A] [IsTopologicalRing A] [T2Space A]
+
+lemma abs_sq_eq_cfcₙ_norm_sq_complex (a : A) (ha : IsStarNormal a) :
+    abs a ^ (2 : NNReal) = cfcₙ (fun z : ℂ ↦ (‖z‖ ^ 2 : ℂ)) a := by
+  conv_lhs => rw [abs_nnrpow_two, ← cfcₙ_id' ℂ a, ← cfcₙ_star, ← cfcₙ_mul ..]
+  exact cfcₙ_congr fun x hx ↦ Complex.conj_mul' x
```

Comment:
optional, just noting that this works here. (because these functions are equal *everywhere*, so we don't need the full power of `cfcₙ_congr` which only requires equality on the spectrum.)
```suggestion
  simp [Complex.conj_mul']
```